Behringer CMD STUDIO 4a mapping.

GandalfMixxx opened the pull request:
This is my new mapping for the Behringer CMD STUDIO 4a.

Forum: http://www.mixxx.org/forums/viewtopic.php?f=7&t=7868
Wiki: http://www.mixxx.org/wiki/doku.php/behringer_cmd_studio_4a

I think it's ready to be included into a Mixxx release, but please let me know if anything needs to be changed.

Regards,
Craig.

GandalfMixxx committed:
BehringerCMDStudio4a mapping V1.2
new file:   res/controllers/Behringer CMDStudio4a.midi.xml
new file:   res/controllers/Behringer-CMDStudio4a-scripts.js

GandalfMixxx committed:
BehringerCMDStudio4a V1.2 mapping.
Added JSHint header as described in the Wiki.

GandalfMixxx committed:
BehringerCMDStudio4a v1.3 mapping.
Update to the mapping to include reverse and reverse-slip plaback,
plus general clean-up of files to comply with coding standards.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+    BehringerCMDStudio4a.initLEDs();
+
+    // Disconnect the VUMeters.
+    // Maybe not! This doesn't seem to work here!
```
What do you mean it doesn't work? What are you trying to accomplish? Generally, the shutdown function should turn off all LEDs.

GandalfMixxx replied:
I do restore the LEDs to their default illumination in initLEDs(), but in init() I have to connect the VU meters to the VuMeterL and VuMeterR parameters with connectControl(). I noticed that to disconnect a control you have to pass an additional true param to the call and assumed that since the script had connected the control in init(), it was the scripts responsibility to clean up and disconnect the control in shutdown(). However, when you try to do this an error is thrown, so I guess this isn't the case. I thus commented out the disconnect (but put a reminder in the code in case I forgot why later, or if anyone else not familiar with Mixxx was wondering about this).

I didn't see anywhere in the Wiki that said you don't have to do this clean-up, and the assumption is (generally) that if you allocate a resource you should deallocate that resource when you are finished with it.

Be-ing commented:
Thanks for submitting this :) I'll take a look through it

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+
+// Function to deal with the DEL "shift" buttons.
+BehringerCMDStudio4a.del = function (channel, control, value, status, group) {
+    BehringerCMDStudio4a.delState[channel]=!BehringerCMDStudio4a.delState[channel];
```
Please put spaces around assignment operators throughout the script for readability. For example:
``BehringerCMDStudio4a.delState[channel] = !BehringerCMDStudio4a.delState[channel];``

GandalfMixxx replied:
Yep, fair point, I do it most of the time, but have missed a few, bad habit! :-) Fixed now.

Be-ing commented:
I think it would make more sense to move the gain control to the FX knobs closest to the mixer.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+
+// Button push/release state variables.
+BehringerCMDStudio4a.pitchDec = {0:false,1:false,2:false,3:false};
+BehringerCMDStudio4a.pitchInc = {0:false,1:false,2:false,3:false};
```
It would make more sense to declare these variables as arrays. For example:
``BehringerCMDStudio4a.pitchInc = [false,false,false,false];

GandalfMixxx replied:
Yep, fixed that. The objects (defined somewhat differently) were a left-over from an early (stupid) version of the code and I forgot about them when I moved to numeric indexes.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+            engine.setValue(group, "play", 1);
+        } else {
+            engine.setValue(group, "play", 0);
+        }
```
You can simplify this function with the recently added script.toggleControl function. Sorry that wasn't documented on the wiki before. Updating http://mixxx.org/wiki/doku.php/midi_scripting#available_common_functions with everything recently added to common-controller-scripts.js is on my to-do list.

You can also simply this by mapping both the push and release of play to the same function.

GandalfMixxx replied:
Ahh! OK, didn't know about the toggleControl() function. I've replaced the conditional toggle with that function now.

However, the push and release functions do different things depending on the mode, i.e. "toggle play" in normal mode, but "hold reverse-play" in  DELETE-mode, (and in fact the release function only has a defined behaviour in DELETE-mode), so I'm not convinced it's a simplification to use the same function for both (then have to add an extra if condition in the "super-function" to determine if the button has been pressed or released).

We have the ability to call different functions for press and release in the XML when the function behaviour for press and release is different, so we should use that ability to simplify the code right?

"Shift/mode" functionality is so fundamental to many controller mappings it seems odd that there is no native support for this in Mixxx at the moment, (i.e. support where you can just define the shift/mode behaviour in the XML with no additional scripting).

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+    } else {
+        BehringerCMDStudio4a.scratchState[channel] = true;
+        midi.sendShortMsg(status, control, 0x01);
+    }
```
This can be simplified to not need an ``if``...``else`` statement by inverting the value of BehringerCMDStudio4a.scratchState[channel] like line 89 and using the ``?`` operator like on line 90.

GandalfMixxx replied:
Yep, well spotted, Fixed.

It was actually like this originally, but then I experimented with a 3 state scratch button:
    normal = MIDI 0x00 = orange button
    scratch = MIDI 0x01 = blue button.
    scratch+slip = MIDI 0x02 = flashing blue button.
I had to use if conditions for that, but the idea didn't pan out, and when I reverted I (lazily) left the if condition in place. My bad!

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+    }
+}
+
+// Functions to deal with the hot-que buttons.
```
typo: hotcue

GandalfMixxx replied:
Well spotted! fixed.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+// Functions to deal with the hot-que buttons.
+BehringerCMDStudio4a.hotcuePush = function (channel, control, value, status, group) {
+    // Translate the button to the actual hotcue.
+    var hotcue = control-0x21;  // Hotcue buttons on left ddeck go from 0x22 to 0x29
```
typo: extra letter "d"

GandalfMixxx replied:
Fixed, thanks again! :-)

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+        // Right deck, buttons are 0x20 higher so we need to compensate.
+        hotcue = hotcue-0x20;
+    }
+    print(hotcue)
```
I'm guessing you used this debug message to develop the above couple lines. At this point it'd probably be best to delete it.

GandalfMixxx replied:
Oops, thought I caught all of those! Fixed :-)

Be-ing commented:
Do the main output and headphone volume knobs control the controller's sound card? If so, mention it on the wiki so users don't wonder why it isn't changing the software gains in Mixxx.

Be-ing commented:
Could you add controls for fixed beat loops (referred to on some controllers as "autoloops")? I suggest making DELETE + Loop on/off toggle a 4 beat loop and DELETE + the arrow buttons halve/double the loop size.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+    } else {
+        engine.setValue(group, "cue_default", 0);
+	}
+}
```
Likewise, this could be simplified by combining both the press and release functions into one.

Be-ing replied:
It looks like you've done this for every button handled by JS. Could you use one function per button throughout?

GandalfMixxx replied:
See earlier comment.
I've only done this where the function is different for each (press or release). I didn't know about the toggleControl()  function at the time, so I could replace the cuePush() and cueRelease() functions with a cue() "super-function" using toggleControl() but thinking about it now, what if the behaviour get's out of sync for some reason (e.g. if the GUI is touched perhaps?, ,or because of some other odd interaction??). If that happens the Push/Release functions would get swapped, (with no simple way for the user to restore normal behaviour). Perhaps this is unlikely, but I believe in "defensive programming" :-)

Perhaps you could get round this by keeping the explicit setValue() calls and wrapping the super function in an if condition to detect whether we are processing a push or a release, but then we are back to adding an if clause to the "super-function" to do this (i.e. creating a more complex function) when we already have the ability to do this test in the XML.

If single functions for both press and release is the Mixxx "house-style" then I will change this (let me know), but (IMHO) it seems simpler the way it is (although again, if shift/mode functionality could be declared directly in the XML without additional scripting, life would be much easier" :-))

Be-ing replied:
I do think that is really unlikely and not really the job of a mapping to deal with. The most likely scenario I could think of in which this might possibly be an issue would be someone holding the cue button on one controller and accidentally hitting it on this controller before releasing it.

Be-ing replied:
Single functions for press and release are the conventional way.

Be-ing replied:
Actually, for this button. script.toggleControl() wouldn't be the best way to do it. Just use ``(value == 0x7f) ? 1 : 0`` as the third parameter of engine.setValue()

GandalfMixxx replied:
OK, I've replaced all the push/release functions with single functions using (value === 127) ? 1 : 0
Code density up, but code readability down I think. :-(

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+BehringerCMDStudio4a.wheelTurn = function (channel, control, value, status, group) {
+    var deck=channel+1;
+    var newValue=value-64;
+    if (BehringerCMDStudio4a.scratchState[channel]){
```
Rename this variable to BehringerCMDStudio4a.scratchButtonState for clarity

GandalfMixxx replied:
OK, done. I've also renamed "delState" to "delButtonState" for consistency.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,257 @@
+        }
+    } else {
+        // DEL mode is not active, just do "rate" changes.
+        engine.setValue(group, "rate_perm_down", 1);
```
Use the rate_temp_down/up controls throughout the script. There's no point in having buttons for rate_perm_down/up when there is a pitch fader.

GandalfMixxx replied:
I did originally have _temp mapping instead of _perm, but changed it for the following reasons.

1) The pitch-fader is limited to +-10% and there have been a few times when I wanted a little bit more than this. Tapping the buttons when they are mapped to _perm gives you this option when the slider is already maxed out.

2) There is no "dead-zone" option for the faders, (and on the CMD STUDIO 4a there are no "zero detents" on the faders either), so it can be *really* hard to return to zero. Using the +/- (_perm) steps instead of the fader makes this precise control easy.
Although this lack of a "zero dead-zone" was such a problem for me that I implemented an additional "zero" function by hitting both + and - together. Perhaps an adjustable "zero dead-zone" option could be added to continuous controllers in a future release??

3) "_temp" functionality is not removed by mapping to _perm. When I need a temporary increase/decrease in rate now, I just tap +/-, then tap the opposite button to cancel, i.e. I now have both _temp (albeit with an extra button tap), and _perm functionality now. This is not true the other way round, i.e. mapping to _temp means I lose all _perm stepping functionality.

If this is a blocker to acceptance, let me know and I will change it, but (IMHO) I think the current behaviour is better.

Be-ing replied:
1. The pitch fader range is configurable in Mixxx's preferences. If you find it too small, that's where it should be handled.
2. That's annoying. An adjustable "zero dead-zone" option seems like it would be really tricky to implement and would involve some tradeoff similar to soft takeover. I suspect it would be better to handle in C++, but if you want to give it a try in JS, go ahead.
3. The rate_temp_down/up controls are meant to only be active while the button is held down. Their purpose is shifting the phase of one track relative to another, not changing the tempo. This is how it is done in the skins.

GandalfMixxx replied:
I'd respond:

1) 10% is fine most of the time, but when you need that little bit extra it's a real pain not to have it at your fingertips, (yep. the range can be adjusted in the prefs, but then you loose precision for the majority of the time you don't need the extra range, and there is *always* going to be a time when you need "a little bit more" no matter what you set the range to).

2) Yep, zero-ing out a rate fader is a pain at the moment, but the + and - together option that I've already implemented works well enough right now. As a feature request, a customisable dead-zone (as you have with most joystick settings) really needs to go in the CC++ code I think, and then be implemented as an XML option, so I won't be trying it in JS :-).

3) Yep, I know what the "_temp" buttons are for, but your suggestion is *not* how it's done in the skins, (at least not the "LateNight" skin I'm using as a guide).
The "LateNight" skin actually has this "_perm" step functionality set-up as the +/- buttons directly under the rate faders. The "_temp" functions are actually mapped to left and right arrows lower down (under these + and - buttons), and as you can see, the CMD STUDIO 4a has +/- buttons directly under the rate faders (not arrows) so my mapping seems much more consistent with the skin-UI than your suggestion.
In any case, my point in item (3) is that you get all the "_temp" functionality you need with "_perm" mapping, plus extra benefits. Instead of holding the button down for a period to change the rate, then releasing the button to return the rate to normal (resulting in a "phase" change), with a "_perm" mapping you now tap the button to change the rate, then tap the opposite button at the end of the period to return the rate to normal. i.e. you still get the phase change, we are only exchanging a button *release* for a button *tap*. In use, you don't lose any "_temp" functionality, *and* you gain all the other benefits with this system (greater range when you need it and repeatable precision step control of rate including zero).

I really think this is a *much* better use of the deck +/- rate buttons. However, if the "_perm" button mapping really is a blocker to acceptance, let me know and I'll change it.

Be-ing replied:
I'm not terribly strongly opposed to keeping it with _perm. I think a third opinion would be helpful here.

Be-ing commented:
Overall the mapping and wiki look pretty good, so thanks for the easy review. :) I edited the wiki a bit. Most changes to the wiki were just small rewordings.

sblaisot commented:
Please add the new files to windows uninstaller
http://www.mixxx.org/wiki/doku.php/contributing_mappings#windows_installer_update

GandalfMixxx commented:
Many thanks for the helpful comments and suggestions. I'll get to work on the changes this evening.

GandalfMixxx commented:
Be-ing commented:
> Do the main output and headphone volume knobs control the controller's sound card? If so,
> mention it on the wiki so users don't wonder why it isn't changing the software gains in Mixxx.

The main and headphone volume knobs don't control the sound card, they do control the software gains in Mixxx ([Master].gain and [Master].headMx) they are set-up in the XML (and do change the software gains in Mixxx properly).

GandalfMixxx commented:
sblaisot commented a day ago:
>Please add the new files to windows uninstaller
>http://www.mixxx.org/wiki/doku.php/contributing_mappings#windows_installer_update

Opps, sorry, completely missed that, fixed now!

GandalfMixxx commented:
Be-ing commented 2 days ago:
> Overall the mapping and wiki look pretty good, so thanks for the easy review. :) I edited the wiki a
> bit. Most changes to the wiki were just small rewordings.

Many thanks, much appreciated.

GandalfMixxx commented:
Be-ing commented:
> I think it would make more sense to move the gain control to the FX knobs closest to the mixer.
> EDIT: Nevermind. With the numbered buttons below it would be a bit awkward that way if one
> knob was above button 1 but the other was above 4.

Yep, the deck layouts aren't very consistent. The controls are mirrored left/right, but the numeric identifiers (the hot-cue and FX-control/assign buttons) aren't mirrored (for obvious reasons) so the layout ends up a bit of a mishmash between the two options.

GandalfMixxx committed:
BehringerCMDStudio4a mapping v1.3
A number of minor (non functional) changes as a reult of the pull request,
plus a few other (non functional) improvements, e.g. Loop start/end release
XML definitions, (so as not to corrupt the GUI state).

GandalfMixxx commented:
Be-ing commented:
> Could you add controls for fixed beat loops (referred to on some controllers as "autoloops")? I
> suggest making DELETE + Loop on/off toggle a 4 beat loop and DELETE + the arrow buttons
> halve/double the loop size. Do you have any ideas for ways to add in controls for moving loops?
> EDIT: Thinking about it more, I think that should be the default way of using the loop controls and
> the manual setting of the in & out points should be the functions activated by holding DELETE.

Mmm..., Yep, auto-loops are missing from the deck layout and I'm not happy about that. I have some ideas about how to squeeze this into the deck, but I need to work out what will work best. Let me try some options out and I'll get back to you.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,247 @@
+        engine.setValue(group, "reverseroll", 1);
+    } else {
+        // DEL-mode is not active, just play.
+		script.toggleControl(group,"play")
```
extra indentation here

GandalfMixxx replied:
Sorry, tabs crept into the file, fixed now.

Be-ing commented on res/controllers/Behringer-CMDStudio4a-scripts.js:
```diff
@@ -0,0 +1,247 @@
+}
+BehringerCMDStudio4a.cueRelease = function (channel, control, value, status, group) {
+    if (BehringerCMDStudio4a.delButtonState[channel]) {
+        // DEL-modde is active, do reverse-roll release.
```
typo: extra "d"

GandalfMixxx replied:
Keyboard seem to be bouncing lately! :-) Fixed.

Be-ing commented:
Okay, leave a comment when you push a new commit.

GandalfMixxx committed:
BehringerCMDStudio4a mapping v1.2
Further (hoppefully non functional) changes to comply with the pull request review.
Rationalised button push/release functions into single button functions.
Minor typos corrected.
Auto-loop functionality is still missing.

GandalfMixxx committed:
BehringerCMDStudio4a mapping v1.4
Added auto-loop functionality.
Added FX change functionality.
DEL button now has to be held to delete a hot cue.
Further code clean-up for pull-request.

GandalfMixxx commented:
New version of the mapping committed. Now includes auto-loop functionality as requested.

GandalfMixxx commented:
Wiki has been updated to reflect the functionality changes in this new version (I've also updated the forum thread).

GandalfMixxx committed:
BehringerCMDStudio4a mapping v1.4
Minor typos corrected.
Changed the slip "re-sync" timer after the discussion in
     https://bugs.launchpad.net/mixxx/+bug/1538200

GandalfMixxx committed:
BehringerCMDStudio4a mapping v1.4
Consolodated the pitch inc/dec functions.

GandalfMixxx committed:
BehringerCMDStudio4a mapping v1.4
Further minor clean-up.

Be-ing commented:
> The main and headphone volume knobs don't control the sound card, they do control the software gains in Mixxx ([Master].gain and [Master].headMx) they are set-up in the XML (and do change the software gains in Mixxx properly).

Do volume controls show up in your OS mixer program when you select the CMD 4a's sound card? If so, these should be used rather than the software gains in Mixxx.

Be-ing commented:
I thought of a few possibilities for the loop button behavior:
* Exit the loop on button up for short loops <= 1 beat, which is the top row of buttons, but not for longer loops in the second row of buttons. I just tried this in the Hercules P32 mapping and I think it's intuitive to use.
* Exit the loop on button up only when slip_enabled is 1.
* Make releasing loops on button up an option that can be easily changed by users by editing the value of a boolean at the top of the script. Or make one of the behaviors above an option.

Be-ing commented:
Sorry for putting this on hold for so long. I got a bit overwhelmed when 5 mappings were submitted in about a week.

Be-ing commented:
@GandalfMixxx , thoughts about the above questions?

Be-ing commented:
@GandalfMixxx : ping. What do you think about the other ways of mapping the loop buttons I suggested above?

rryan commented:
It would be great to have this in Mixxx 2.1 -- @GandalfMixxx @Be-ing what's left to do here?

GandalfMixxx commented:
When I last worked on this about a year ago it was ready for inclusion as far as I was concerned. I'd addressed all the comments/suggestions from Be-ing and also created what I considered to be pretty good documentation on the Wiki (I'd also posted everything to the Mixxx forum for others to try out). Then I didn't get any further replies for months and still no merge so I moved on to other things.

Three months later Be-ing pops up with more suggestions for alterations that didn't make any real sense to me and I really didn't have the energy to come back to this only to get more objections to a merge. I figured that the mapping is out there and easily downloadable for those that actually have the controller (from the forum), and the documentation is on the Wiki too. If it's not good enough for a merge then fair enough.

More months pass, then in November another post from Be-ing, again not talking about a merge but instead asking me to go over his suggestions from over 6 months ago??!! Sorry, I made every effort to give back to the community already and got nowhere. My time is too valuable to mess about like this.


So here we are a year later... All I can say is that a year ago the mapping was working well enough for me. It was (as far as I was concerned) feature complete, and while it may not be perfect, or suit everyone, I think it was good enough for inclusion back then (even if only on the basis of a first version that could be extended later). I can't honestly say if it will work as well with the latest version of Mixxx as I've not installed it, but I see that there have been some downloads from the forum, some generally positive feedback, and some useful modifications from people that actually own (or owned) the controller (that I might have taken the time to roll into the mapping here if my experience with this contribution hadn't been so negative).

I'm happy for you to include it in a version of Mixxx "as is" if you are now happy with it or for you to reject the PR if you think it's still not up to scratch. I may come back to work on this mapping in the future but it probably won't be for a while yet as I'm busy with other things right now.

Be-ing commented:
It would be good to update the mapping for the new effects UI, but I'd be okay with merging it as is. I think there's room for improvement with how the loop buttons are mapped, but it's not bad how it is.

Be-ing commented:
I apologize for not responding for a long time when you first submitted this. Coincidentally when you submitted this, 4 other mappings were submitted within a week so I got overwhelmed.

Be-ing commented:
Thank you for your contribution. Hopefully the user who posted a modified version on the forum will want to take up the work of updating it for the new effects UI.

rryan commented:
Sounds good -- let's merge this then. @GandalfMixxx thanks for your contribution and sorry for the delay.

Be-ing commented:
This was targeted at the 1.12 branch. @rryan can you merge it into master?

rryan commented:
> This was targeted at the 1.12 branch. @rryan can you merge it into master?

Good catch, done.
https://github.com/mixxxdj/mixxx/commit/f3905bab698a9cedadab96ed85cb36697e6035ff
